### PR TITLE
Modernize CashAddress code

### PR DESCRIPTION
**Pull request by Milerius**
## Description

- Use inline variable instead of static const variable when possible
- noexcept
- refactor to respect DRY idiom
- use `nodiscard` for functions that should use their result
- do not use `strncmp` - use `string::compare` instead
- use braces initializer to initialize a variable
- use for the prefix a version and a switch case instead of if/elseif
- remove unused code

## How to test

```bash
make -C build tests && build/tests/tests tests
```

## Types of changes

Modernize code (use modern c++, improve readability, more safeness)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add/Update tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] If there is a related Issue, mention it in the description.

### CODE REVIEW

File: src/Bitcoin/Amount.h [hunk 1]
```diff
@@ -16,14 +16,9 @@ namespace TW::Bitcoin {
...
-static const Amount coin = 100000000;
+inline constexpr Amount coin = 100000000;
 
 /// Maxximum valid amount in satoshis.
-static const Amount maxAmount = 21000000 * coin;
-
-/// Detemines if the provided value is a valid amount.
-inline bool isValidAmount(const Amount& amount) {
-    return (amount >= 0 && amount <= maxAmount);
-}
```
Comment (Milerius) [inline src/Bitcoin/Amount.h:27]: was unused, I removed it

File: src/Bitcoin/CashAddress.h [hunk 2]
```diff
@@ -7,22 +7,27 @@
...
+
+#include <TrustWalletCore/TWHRP.h>
 
 #include <cstdint>
 #include <string>
 
 namespace TW::Bitcoin {
 
+inline const std::string gBitcoinCashHrp{HRP_BITCOINCASH};
+inline const std::string gECashHrp{HRP_ECASH};
```
Comment (Milerius) [inline src/Bitcoin/CashAddress.h:21]: It's better than static const variable member initialization, but require an additional include as showed above

**Review by Milerius — COMMENTED** [review]
I've also deleted some constructor that were not used at all, please let me know if we should keep them

File: src/Bitcoin/CashAddress.cpp [hunk 3]
```diff
@@ -12,70 +12,74 @@
...
-    }
-    std::copy(data.begin(), data.begin() + dataLen, bytes.begin());
+    const bool decodeValid =
+        cash_decode(decodedHRP.data(), data.data(), &dataLen, withPrefix.c_str()) != 0 &&
+        dataLen == CashAddress::size;
+    return decodeValid &&
+           std::string(decodedHRP.data()).compare(0, std::min(hrp.size(), maxHRPSize), hrp) == 0;
 }
 
-CashAddress::CashAddress(const std::string& hrp, const Data& data)
```
[the remark this answers is not in the document]
Comment (catenocrypt) [reply src/Bitcoin/CashAddress.cpp:65]: Apparently it was not used

[the remark this answers is not in the document]
Comment (Milerius) [reply src/Bitcoin/CashAddress.cpp:65]: > Apparently it was not used

Yes

**Review by hewigovens — APPROVED** [review]

**Review by catenocrypt — APPROVED** [review]

**Comment by SORA-ALONE** [comment]
Milerius:dev_roman #2259

**Pull request merged**